File: common/src/main/java/cn/breadnicecat/candycraftce/block/blocks/CandyWaterPlantBlock.java
package cn.breadnicecat.candycraftce.block.blocks;

import net.minecraft.core.BlockPos;
import net.minecraft.tags.FluidTags;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.block.LiquidBlockContainer;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.material.Fluid;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.material.Fluids;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Created in 2024/2/14 23:32
 * Project: candycraftce
 *
 * @author <a href="https://github.com/BreadNiceCat">Bread_NiceCat</a>
 * <p>
 */
public class CandyWaterPlantBlock extends CandyPlantBlock implements LiquidBlockContainer {
	protected CandyWaterPlantBlock(Properties properties) {
		super(properties);
	}

	@Override
	public boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
		return state.getFluidState().is(FluidTags.WATER) && super.canSurvive(state, level, pos);
	}


	@Nullable
	public BlockState getStateForPlacement(BlockPlaceContext context) {
		FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
		return fluidState.is(Fluids.WATER) ? super.getStateForPlacement(context) : null;
	}

	@SuppressWarnings("deprecation")
	public @NotNull FluidState getFluidState(BlockState state) {
		return Fluids.WATER.getSource(false);
	}

	@Override
	public boolean canPlaceLiquid(BlockGetter level, BlockPos pos, BlockState state, Fluid fluid) {
		return false;
	}

	@Override
	public boolean placeLiquid(LevelAccessor level, BlockPos pos, BlockState state, FluidState fluidState) {
		return false;
	}
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/core/block/blocks/plant/CandyWaterPlantBlock.kt
package cn.breadnicecat.candycraftce.core.block.blocks.plant

import net.minecraft.core.BlockPos
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.LevelAccessor
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.block.LiquidBlockContainer
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.material.Fluid
import net.minecraft.world.level.material.FluidState
import net.minecraft.world.level.material.Fluids

/**
 * Created by NiceCat on 2025/12/5.
 * Project: candycraftce
 * <p>
 *
 */
class CandyWaterPlantBlock(properties: Properties) : CandyPlantBlock(properties), LiquidBlockContainer {
    @Deprecated("Deprecated in Java")
    @Suppress("DEPRECATION")
    override fun canSurvive(state: BlockState, level: LevelReader, pos: BlockPos): Boolean {
        return state.fluidState.`is`(Fluids.WATER)
                && super.canSurvive(state, level, pos)
    }


    override fun getStateForPlacement(context: BlockPlaceContext): BlockState? {
        val fluidState = context.level.getFluidState(context.clickedPos)
        return if (fluidState.`is`(Fluids.WATER)) super.getStateForPlacement(context) else null
    }

    @Deprecated("Deprecated in Java")
    override fun getFluidState(state: BlockState): FluidState =
        Fluids.WATER.getSource(false)

    override fun canPlaceLiquid(
        level: BlockGetter,
        pos: BlockPos,
        state: BlockState,
        fluid: Fluid,
    ): Boolean = false

    override fun placeLiquid(
        level: LevelAccessor,
        pos: BlockPos,
        state: BlockState,
        fluidState: FluidState,
    ): Boolean = false

}